### ISSUE TITLE
[❔ other question] How to deal with light source occlusion by a glass plate

### ISSUE THREAD

**Issue by PidgeyBE** [mitsuba-renderer/mitsuba2#262]
## Summary

In the docs I read https://mitsuba2.readthedocs.io/en/latest/generated/plugins.html#path-tracer-path
> This directly translates into a category of scenes where a path tracer can be expected to produce reasonable results: this is the case when the emitters are easily “accessible” by the contents of the scene. For instance, **an interior scene that is lit by an area light will be considerably harder to render when this area light is inside a glass enclosure (which effectively counts as an occluder)**.

Is there a way to overcome this? How do I render such a case?
## Description
I'm trying to render a scene like the one below:
![image](https://user-images.githubusercontent.com/19860056/91012154-ea560980-e5e5-11ea-858e-78331b311404.png)
To make sure no stray light is influencing the rendering, I've put my light source in a black tube.
However, if I place a glass plate **halfway** in between the light source and the projected image, I get:
![image](https://user-images.githubusercontent.com/19860056/91012261-17a2b780-e5e6-11ea-946d-19994c4c4725.png)
In the image we see that the glass plate is indeed transparent to light that is not comming directly from the light source, but the glass plate seems to be opaque to light rays comming directly from the light source. Is there a workaround so light is not blocked by transparent materials?

## Steps to reproduce
As the situation is also mentionned in the docs, I assume this is a known issue with a hopefully known solution :grimacing: 

Thanks a lot for any help! Pieterjan

**Comment by Speierers** [comment]
Hi @PidgeyBE ,

This case is particularly difficult for the path tracer as it won't be able to perform next event estimation accross the glass plate. This means that it needs to find the light "by chance" as it constructs the light paths.

Is the glass plate perpendicular to the incoming light in this scenario? It looks to me like if the orientation of the glass plate refacts the light with a different "perspective".

It would be helpful to project a grid pattern instead of this image in order to better understand what's going on.

BTW, placing the black tube around the light will likely increase the variance of the output image as many rays will hit the tube when trying to reach the light. Which emitter are you using here?

**Comment by PidgeyBE** [comment]
Hi @Speierers ,

Thanks for your help!
The emiter I'm using is this:
```
<shape type="obj">
	<string name="filename" value="rectangle.obj"/>
	<transform name="toWorld">
	    <scale x="200" y="200"/>
            <rotate y="1" angle="0"/>
            <translate x="0" z="-1000" y="0"/>
	</transform>
        <ref id="light"/>
	<emitter type="area">
		<spectrum name="radiance" value="400:2000, 700:2000"/>
	</emitter>
</shape>
```
The reason for the black tube is also to filter out angled light rays, so I'm trying to obtain a more or less collimated light source.
The glass plate is in a plane parallel to the light source, so most rays should be perpendicular to the glass plate.
Is there a better way to have a ~collimated light source?

**Comment by Speierers** [comment]
Hi @PidgeyBE ,

Sounds to me like the `projector` emitter plugin is what you are looking for here. It is basically the equivalent of a perspective camera for emitters. This way you should be able to get ride of the black tube, which should reduce the variance in your image a bit.

Could you make sure the normals of your glass plate are correct? Also what is the IOR of the glass plate in your scene?

**Comment by PidgeyBE** [comment]
Hi @Speierers 

I didn't understand how I could have overlooked the `projector` plugin, but now I see.
I'm on the `pathreparam-optix7` branch, because my final goal is to optimize the glass slab surface just like in the mitsuba2 paper. But I see this branch is lagging behind on master and does not have the projector plugin yet. I tried to merge them but there are too many merge conflicts I'm not comfortable with :grimacing: .
So I'll be a little more patient and wait until the branches are in sync :+1: 

The normals of my glass plate are correct. I've tried different plates and checked the normals in Blender, so I'm quite sure. IOR is 1.5.

**Comment by Speierers** [comment]
Could you do a couple renders with different IOR values to check whether this behave properly? I am not even sure the image above is incorrect.

**Comment by merlinND** [comment]
@PidgeyBE Note that results from the Mitsuba 2 paper (including the glass surface optimizations) did not use `pathreparam`. #196 is quite related to what you are doing, perhaps the discussion there will help.

**Comment by PidgeyBE** [comment]
Hey @Speierers 
I did a couple of renders with IORs: 1 - 1.33 - 1.5 - 2 
![image](https://user-images.githubusercontent.com/19860056/92307480-2484bb00-ef97-11ea-8508-e41327c54959.png)  ![image](https://user-images.githubusercontent.com/19860056/92307481-28b0d880-ef97-11ea-96a9-818e9b71fcb2.png)  ![image](https://user-images.githubusercontent.com/19860056/92307483-2babc900-ef97-11ea-8fed-b9c5ece9ae22.png)  ![image](https://user-images.githubusercontent.com/19860056/92307486-2f3f5000-ef97-11ea-8b4e-b1ba07b2ef1b.png)
The (first) render with IOR = 1 shows what is to be expected I guess? There are no reflections at the edges of the glass plate, but there is a black shadow behind the glas plate because of the way the integrator/path tracer works?

@merlinND Aha! I thought the results from the Mitsuba2 paper were indeed using `pathreparam`. I read #196 and indeed, I'm trying to accomplish/reproduce the same. It's a pity/confusing it hasn't been ported to mitsuba2 (yet) as the paper is about mitsuba2, but I know how software development goes :D Thanks for all the other efforts!

So to summarize, I will not be able to use mitsuba2 in it's current state (master/pathreparam) to optimize the surface of a glass plate, similar to what is published paper?

**Comment by merlinND** [comment]
@PidgeyBE To reproduce the results from the paper, the following two plugins are missing: a light tracer and a collimated area light source (see #196).
